### PR TITLE
Adjust order limit prices for volume based fee

### PR DESCRIPTION
**Pull request by sunce86**
# Description
Fixes https://github.com/cowprotocol/services/issues/2503

# Changes
Limit prices are adjusted after being scaled (for partial fillable orders) so that the order limit prices sent to solvers are WORSE than real ones, so that the driver has more room to withhold the fee on solution without violating limit prices.

## How to test
Existing tests.
e2e tests `local_node_volume_fee_sell_order` and `local_node_volume_fee_buy_order`. No change in the expectation which is good. Observed locally that we actually send adjusted sell/buy limit amounts to solvers but the solution is the same, since it does not affect the baseline solver.

For driver tests, expectation on what limit amounts are sent from driver to solver are changed, which is expected.

### CODE REVIEW

File: crates/driver/src/domain/competition/order/mod.rs [hunk 1]
```diff
@@ -190,6 +190,28 @@ impl Order {
                 .unwrap_or_default()
                 .into();
 
+        // adjust ammounts if order has volume based protocol fee
```
Comment (fleupold) [inline crates/driver/src/domain/competition/order/mod.rs:193]: Can you reference the issue here that will allow solvers to skip all driver fee adjustments (so we can more easily find the places that need adjustments). 

Also are you sure this is the right place to add this logic? Won't available balance smaller than the full balance cause issues for fill or kill orders (an e2e test should tell us)?

File: crates/driver/src/domain/competition/order/mod.rs [hunk 2]
```diff
@@ -190,6 +190,28 @@ impl Order {
...
+                    amounts.sell.amount -= amounts
+                        .sell
+                        .amount
+                        .apply_factor(*factor)
+                        .unwrap_or_default();
+                }
+                Side::Sell => {
+                    // increase buy amount by factor
+                    amounts.buy.amount +=
+                        amounts.buy.amount.apply_factor(*factor).unwrap_or_default();
```
Comment (fleupold) [inline crates/driver/src/domain/competition/order/mod.rs:211]: This seems unnecessarily inaccurate. If the orders buy amount is 100 and there is a 1% volume fee, the adjusted amount shouldn't be 100 + 1% but instead 100 / 0.99, so that the reciprocal operation (taking 1% from the result) is 100.

**Review by fleupold — COMMENTED** [review]
Let's add an e2e test with your currently unsupported case as an acceptance test for merging this PR (should be possible by doing slightly cleaner math)

File: crates/driver/src/domain/competition/order/mod.rs (hunk 1 above)
Comment (sunce86) [reply crates/driver/src/domain/competition/order/mod.rs:193]: > Won't available balance smaller than the full balance cause issues for fill or kill orders (an e2e test should tell us)

If you refer to function `filter_orders` where we filter orders with insufficient balance, then AFAIS this change affects only the BUY orders and it lowers the required sell amount which means filtering is less strict than it was before. Although, truly introduces imprecision so I will move it out. 

If you refer to fulfillments requirement where we require solvers to return executed amount and fees which are in total equal to sell amount for sell orders, I don't think this affects this logic since it works with surplus token and not with target token. 

Will add tests to confirm this.

File: crates/driver/src/domain/competition/order/mod.rs (hunk 2 above)
Comment (sunce86) [reply crates/driver/src/domain/competition/order/mod.rs]: Ah yes 🥇 I took completely wrong path. Fixed so no more edge case.

File: crates/driver/src/domain/competition/order/mod.rs (hunk 1 above)
Comment (sunce86) [reply crates/driver/src/domain/competition/order/mod.rs]: Moved calculation out. Added tests.

**Review by Mateo-mro — APPROVED** [review]

**Review by squadgazzz — APPROVED** [review]
👍
